### ISSUE TITLE
Optimize Models / Hyperparameter Tuning

### ISSUE THREAD

**Issue by xehu** [Watts-Lab/team_comm_tools#156]
Currently, we are not performing model optimization or hyperparameter tuning, even in our baseline models. There is a rather large "TODO" and commented-out code in `model_builder.py`.

Issue https://github.com/Watts-Lab/team-process-map/issues/116 has already discussed model optimization issues in a few ways, but we have not yet acted on them. These ideas are reproduced below.

> Improve Model Performance and Optimization
> The current models perform very poorly in terms of most of the metrics we have used. The 2 main hypothesis on why this might be the case are:
> We are not choosing the correct proxy for performance across each dataset. i.e. the targets are not easy/appropriate to model.
> We are not tuning the model well.
> This work-stream would aim to explore different combinations of targets to see whether there is signal in the dataset organically, and also implement a more robust, scalable and effective hyperparameter tuning mechanism to extract the best performance from our baseline models.

> 3. improving_parameter_tuning
> Currently we are running bayesian optimization as a way to search the grid of hyperparameters automatically, and come up with a reasonable setting in the least amount of trials. We ideally would want to come up with some kind of human in the loop approach where we deduce the model performance characteristics of the baseline model (is it overfitting or under-fitting) based on better model diagnostics visuals (better and more informative SHAP scores etc.), and then choose the parameter space that makes the most sense to explore.
> NOTE: Given that we plan to tackle optimization more rigorously in a future PR (implementation will change dramatically based on the model architecture we finalize from the feature_bagging_eda branch), in the baseline model (https://github.com/Watts-Lab/team-process-map/pull/112) we have commented out the naive way of parameter optimization for now.

# Questions / Action Items
- In order to properly tune hyperparameters, we need a **proper held-out test set**, one that doesn't come from our main data, and that we can use to test our models after training them (on the training set) and tuning hyperparameters (on the validation set). Currently, the `train-test-split` function in the ModelBuilder simply splits the data into three sections, but does not allow for (1) hyperparameter tuning on the validation set, as the model is simply evaluated twice; nor (2) creates a true held-out test set, as it is simply a random split created from the same data that we use for training and validation.
- Determine which hyperparameters we want to tune for which models. This includes: (1) parameters related to the random forest, e.g., number of tree, split depth; (2) parameters related to feature down-selection, e.g., the correlation coefficient used to drop features during down-selection.
- Implement a separate pipeline for conducting hyperparameter tuning (e.g., grid search).
- Finally, putting the pieces together: Creating the held-out test set --> Implementing the hyperparameter turning/optimization --> Running and evaluating the models on a true train/validation/test split.

**Comment by xehu** [comment]
Closing as this is about a previous version of the project and no longer related to the toolkit
